relay-replays/src/recording.rs

```diff
@@ -231,59 +235,72 @@ impl RecordingProcessor<'_> {
 /// -> META = 4
 /// -> CUSTOM = 5
 /// -> PLUGIN = 6
-
 #[derive(Debug, Serialize)]
 #[serde(untagged)]
-enum Event {
-    T2(FullSnapshotEvent),
-    T3(IncrementalSnapshotEvent),
-    T4(MetaEvent),
-    T5(CustomEvent),
+enum EventData {
+    FullSnapshot(FullSnapshotEventData),
+    IncrementalSnapshot(IncrementalSourceData),
+    Meta(Value),
+    Custom(CustomEventData),
     Default(Value),
-    // 0: DOMContentLoadedEvent,
-    // 1: LoadEvent,
-    // 6: PluginEvent,
+}
+
+#[derive(Debug, Serialize)]
+struct Event {
+    #[serde(rename = "type")]
+    ty: u8,
+    timestamp: f64,
+    data: EventData,
 }
 
 impl<'de> serde::Deserialize<'de> for Event {
     fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
-        let value = Value::deserialize(d)?;
-
-        match value.get("type") {
-            Some(val) => match Value::as_u64(val) {
-                Some(v) => match v {
-                    2 => match FullSnapshotEvent::deserialize(value) {
-                        Ok(event) => Ok(Event::T2(event)),
-                        Err(_) => Err(DError::custom("could not parse snapshot event")),
-                    },
-                    3 => match IncrementalSnapshotEvent::deserialize(value) {
-                        Ok(event) => Ok(Event::T3(event)),
-                        Err(_) => Err(DError::custom("could not parse incremental snapshot event")),
-                    },
-                    4 => match MetaEvent::deserialize(value) {
-                        Ok(event) => Ok(Event::T4(event)),
-                        Err(_) => Err(DError::custom("could not parse meta event")),
-                    },
-                    5 => match CustomEvent::deserialize(value) {
-                        Ok(event) => Ok(Event::T5(event)),
-                        Err(e) => Err(DError::custom(e.to_string())),
-                    },
-                    0 | 1 | 6 => Ok(Event::Default(value)),
-                    _ => Err(DError::custom("invalid type value")),
-                },
-                None => Err(DError::custom("type field must be an integer")),
-            },
-            None => Err(DError::missing_field("type")),
+        #[derive(Deserialize)]
+        #[serde(bound(deserialize = "'de: 'a, 'a: 'de"))]
+        struct Helper<'a> {
+            #[serde(rename = "type")]
+            ty: u8,
+            timestamp: f64,
+            data: Content<'a>,
```

Comment:
The basic idea here is to use serde's `Content` for more efficient deserialization, which is what serde itself uses for [deserializing internally tagged enums](https://cs.github.com/serde-rs/serde/blob/b7be637e8cf73dcbea28c6ea960604371726171d/serde_derive/src/de.rs?q=repo%3Aserde-rs%2Fserde+ContentDeserializer#L1332).

Since `Content` is not public API, we might also try using [`serde_json::RawValue`](https://docs.rs/serde_json/latest/serde_json/value/struct.RawValue.html) instead.

Reply:
Content internally has a secret interface to avoid re-buffering when used recursively. RawValue doesn’t habe that *and* it has the added issue that it uses in-band signalling which would also change the input interface of any incoming json payload.